crates/subspace-service/src/dsn/piece_record_store.rs

```diff
@@ -0,0 +1,125 @@
+use sc_client_api::AuxStore;
+use sc_piece_cache::{AuxPieceCache, PieceCache, MAX_SEGMENTS_NUMBER_IN_CACHE};
+use std::borrow::Cow;
+use std::marker::PhantomData;
+use std::sync::Arc;
+use subspace_core_primitives::{PieceIndex, PieceIndexHash, PIECES_IN_SEGMENT};
+use subspace_networking::libp2p::kad::record::Key;
+use subspace_networking::libp2p::kad::Record;
+use subspace_networking::{RecordStorage, ToMultihash};
+use tracing::{trace, warn};
+
+pub type SegmentIndexGetter = Arc<dyn Fn() -> u64 + Send + Sync + 'static>;
+
+pub struct AuxRecordStorage<AS> {
+    piece_cache: AuxPieceCache<AS>,
```

Comment:
I'm not sure we need piece getter abstraction anymore, you can just incorporate its logic here, no need to have two abstractions I think. With that `last_segment_index_getter` will no longer be necessary either.

Let's have a TODO for that since we'll need DSN to be enabled by default for that.
I think since farmer connects to node regardless, it may as well exchange DSN address information so node connects to the farmer on DSN level and we don't need to expose pieces over RPC anymore and DSN will effectively be enabled all the time, even if it is between node and farmer in isolation from the rest of the network.

Reply:
I don't get how the `last_segment_index_getter` on the node is connected with the piece-getter type on the farmer. The former relates to the absence of the iterator in the AuxStore and the latter receives pieces using different channels. 

Nevertheless, I added a TODO for the RPC piece-getter on the farmer's side.

Reply:
When `AuxPieceCache` is absorbed by `AuxRecordStorage` implementation (`AuxPieceCache` will not be needed otherwise once we remove RPC endpoint), it'll be responsible for storing new pieces and will know exactly what the persisted range of pieces is rather than needing `last_segment_index_getter`.

I was thinking of having a TODO here so we don't forget to prune the unnecessary layers of abstraction later.